src/main/java/edu/ie3/datamodel/io/sink/SqlSink.java

```diff
@@ -0,0 +1,507 @@
+
+    // get raw grid entities with types or operators
+    RawGridElements rawGridElements = jointGridContainer.getRawGrid();
+    Set<NodeInput> nodes = rawGridElements.getNodes();
+    Set<LineInput> lines = rawGridElements.getLines();
+    Set<Transformer2WInput> transformer2Ws = rawGridElements.getTransformer2Ws();
+    Set<Transformer3WInput> transformer3Ws = rawGridElements.getTransformer3Ws();
+    Set<SwitchInput> switches = rawGridElements.getSwitches();
+    Set<MeasurementUnitInput> measurementUnits = rawGridElements.getMeasurementUnits();
+
+    // get system participants with types or operators
+    SystemParticipants systemParticipants = jointGridContainer.getSystemParticipants();
+    Set<BmInput> bmPlants = systemParticipants.getBmPlants();
+    Set<ChpInput> chpPlants = systemParticipants.getChpPlants();
+    Set<EvcsInput> evCS = systemParticipants.getEvcs();
+    Set<EvInput> evs = systemParticipants.getEvs();
+    Set<FixedFeedInInput> fixedFeedIns = systemParticipants.getFixedFeedIns();
+    Set<HpInput> heatPumps = systemParticipants.getHeatPumps();
+    Set<LoadInput> loads = systemParticipants.getLoads();
+    Set<PvInput> pvPlants = systemParticipants.getPvPlants();
+    Set<StorageInput> storages = systemParticipants.getStorages();
+    Set<WecInput> wecPlants = systemParticipants.getWecPlants();
+
+    // get graphic elements (just for better readability, we could also just get them directly
+    // below)
+    GraphicElements graphicElements = jointGridContainer.getGraphics();
+
+    // extract types
+    Set<AssetTypeInput> types =
+        Stream.of(
+                lines,
+                transformer2Ws,
+                transformer3Ws,
+                bmPlants,
+                chpPlants,
+                evs,
+                heatPumps,
+                storages,
+                wecPlants)
+            .flatMap(Collection::stream)
+            .map(Extractor::extractType)
+            .collect(Collectors.toSet());
+
+    // extract operators
+    Set<OperatorInput> operators =
+        Stream.of(
+                nodes,
+                lines,
+                transformer2Ws,
+                transformer3Ws,
+                switches,
+                measurementUnits,
+                bmPlants,
+                chpPlants,
+                evCS,
+                evs,
+                fixedFeedIns,
+                heatPumps,
+                loads,
+                pvPlants,
+                storages,
+                wecPlants)
+            .flatMap(Collection::stream)
+            .map(Extractor::extractOperator)
+            .flatMap(Optional::stream)
+            .collect(Collectors.toSet());
+
+    List<Entity> toAdd = new LinkedList<>();
+    toAdd.addAll(rawGridElements.allEntitiesAsList());
+    toAdd.addAll(systemParticipants.allEntitiesAsList());
+    toAdd.addAll(graphicElements.allEntitiesAsList());
+    toAdd.addAll(types);
```

Comment:
Can't you get all the elements with the e.g. systemParticipants.allEntitiesAsList() method ? The types and operators should be extracted automatically when persisting the participants, don't they?

Reply:
I don't think it will be that easy, because not every system participant has a type. But maybe we can move this to another issue, because csv related it's the same.

Reply:
It does because you handle nested entities inside `persistAll`

```java
  /** Persists a whole {@link JointGridContainer}. */
  public void persistJointGrid(JointGridContainer jointGridContainer, UUID gridUUID) {
    DbGridMetadata identifier = new DbGridMetadata(jointGridContainer.getGridName(), gridUUID);
    List<Entity> toAdd = new LinkedList<>();
    toAdd.addAll(jointGridContainer.allEntitiesAsList());
    persistAll(toAdd, identifier);
  }
```

is sufficient

Reply:
Thank you! This method is now more simple!